maze_dataset/tokenization/maze_tokenizer.py

```diff
@@ -517,19 +653,247 @@ def to_tokens(self, coord: Coord | CoordTup) -> list[str]:
+                sorted_edges,
+                np.unique(sorted_edges[:, 0, :], return_index=True, axis=0)[1][1:],
+            )
+            if self.shuffle_group:
+                [numpy_rng.shuffle(g, axis=0) for g in groups]
+            return groups
+
+
+class EdgePermuters(_TokenizerElementNamespace):
+    """Namespace for `EdgePermuter` subclass hierarchy used by `AdjListTokenizer`."""
+
+    key = "edge_permuter"
+
+    @serializable_dataclass(frozen=True, kw_only=True)
+    class _EdgePermuter(TokenizerElement, abc.ABC):
+        """Specifies how to sequence the two coords that encode a lattice edge."""
+
+        @classmethod
+        def attribute_key(cls) -> str:
+            return EdgePermuters.key
+
+        def is_valid(self) -> bool:
+            # No invalid instances possible within data member type hint bounds
+            return True
+
+        @staticmethod
+        @abc.abstractmethod
+        def _permute(lattice_edges: ConnectionArray) -> ConnectionArray:
+            """
+            Executes a permutation.
+            Warning: Caller should be aware that `lattice_edges` may be modified in-place depending on the subclass's implementation.
+
+            # Parameters
+            - `lattice_edges`: Array of lattice edges.
+            The two coords in shape[1] must be adjacent in the lattice.
+
+            # Returns
+            - Array of lattice edges with entries along shape[1] systematically permuted.
+            - shape[0] of the returned array is NOT guaranteed to match `lattice_edges.shape[1]`.
+            """
+            pass
+
+    @serializable_dataclass(frozen=True, kw_only=True)
+    class RandomCoord(_EdgePermuter):
+        """Permutes randomly."""
+
+        @staticmethod
+        def _permute(lattice_edges: ConnectionList) -> ConnectionArray:
+            # return np.random.shuffle(lattice_edges.transpose((1,0,2,3))).transpose((1,0,2,3))
+            numpy_rng.shuffle(lattice_edges, 1)
+            return lattice_edges
+
+    @serializable_dataclass(frozen=True, kw_only=True)
+    class BothCoords(_EdgePermuter):
+        """Includes both possible permutations of every edge in the output.
+        Since input ConnectionList has only 1 instance of each edge,
+        a call to `BothCoords._permute` will modify `lattice_edges` in-place to double `shape[1]`.
+        """
+
+        @staticmethod
+        def _permute(lattice_edges: ConnectionList) -> ConnectionArray:
+            return np.append(lattice_edges, np.flip(lattice_edges, axis=1), axis=0)
+
+
+class EdgeSubsets(_TokenizerElementNamespace):
+    """
+    Namespace for `EdgeSubset` subclass hierarchy used by `AdjListTokenizer`.
+    """
+
+    key = "edge_subset"
+
+    @serializable_dataclass(frozen=True, kw_only=True)
+    class _EdgeSubset(TokenizerElement, abc.ABC):
+        """
+        Component of an `AdjListTokenizers._AdjListTokenizer` which specifies the subset of lattice edges to be tokenized.
```

Comment:
under what circumstances would we need to only tokenize a subset of the edges? I'm confused here

Reply:
The complete set of lattice edges for a maze is `2*n*(n-1)` long. The legacy default behavior is to only tokenize the subset of those edges which are connections in the maze That's `ConnectionEdges`. But there's another option where you can tokenize all lattice edges and use 1 of 2 tokens to indicate for each edge if it's a connection or a wall. That's `AllLatticeEdges`.